### PR TITLE
Port cmp with no extra slot

### PR DESCRIPTION
**Pull request by nanjekyejoannah**
Due to segfaults introducing a new `tp_compare` slot proved problematic. I have found a way of supporting `cmp` without a new slot. Tests are updated to match the new functionality where Py2.x doesn't fail.

I wanted to force push on [this branch] (https://github.com/softdevteam/pygrate3) but maybe you wanted to compare before I force push.

This replaces #4

### CODE REVIEW

**Comment by vext01** [comment]
Out of interest, why did adding the new slot cause a problem?

**Comment by nanjekyejoannah** [comment]
Adding a new slot requires that we inject a dependency to support the operation for every object in Python, where we dont require `cmp`, we have to add it nevertheless with a `0` while where its required with custom logic e.g `dicts` we attach a routine like `dict_compare` without a slot we can still support `cmp` in the few cases we need it i.e for py2.x compatibility. 

For a clean design, its better to use a slot, but this dependency brings hard to understand and trace bugs, my main one related to frozen modules.

**Comment by vext01** [comment]
Was it not possible to add a dummy `cmp` in the base `object` class? Maybe it doesn't work like that.

**Comment by ltratt** [comment]
This is certainly a much simpler PR, which is hopefully a good sign! Have we ported over all the tests from https://github.com/softdevteam/pygrate3/pull/4 into this PR?

bors try

**Comment by ltratt** [comment]
I'm not sure if those failures are particularly related to this PR or not?

**Comment by nanjekyejoannah** [comment]
> This is certainly a much simpler PR, which is hopefully a good sign! Have we ported over all the tests from #4 into this PR?
> 
> bors try

All relevant tests are added from the previous PR.

**Comment by nanjekyejoannah** [comment]
> I'm not sure if those failures are particularly related to this PR or not?

I cant see the error from the build bots to know what is failing, all browsers I have are failing but I have just verified that no tests fail for the PR.

**Comment by ltratt** [comment]
At least for `test_types` one can see from http://ci.soft-dev.org/#/builders/1/builds/1754/steps/2/logs/stdio that:

```
test test_types failed -- Traceback (most recent call last):
  File "/ci/Lib/test/test_types.py", line 1016, in test_methods
    self.assertEqual(attrs, {
    ^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: Items in the second set but not the first:
'__cmp__'
```

**Comment by nanjekyejoannah** [comment]
> At least for `test_types` one can see from http://ci.soft-dev.org/#/builders/1/builds/1754/steps/2/logs/stdio that:
> 
> ```
> test test_types failed -- Traceback (most recent call last):
>   File "/ci/Lib/test/test_types.py", line 1016, in test_methods
>     self.assertEqual(attrs, {
>     ^^^^^^^^^^^^^^^^^^^^^^^^^
> AssertionError: Items in the second set but not the first:
> '__cmp__'
> ```

Doesn't show for me. 

Anyway lets do bor try and see for this revision.

**Comment by ltratt** [comment]
> Doesn't show for me.

Where are you looking?

**Comment by ltratt** [comment]
bors try

**Comment by nanjekyejoannah** [comment]
> > Doesn't show for me.
> 
> Where are you looking?



> > Doesn't show for me.
> 
> Where are you looking?

I checked that link: https://ci.soft-dev.org/#/builders/1/builds/1754/steps/2/logs/stdio
Says cant load the page

File: Objects/typeobject.c [hunk 1]
```diff
@@ -6841,6 +6841,26 @@ wrap_objobjargproc(PyObject *self, PyObject *args, void *wrapped)
     Py_RETURN_NONE;
 }
 
+static PyObject *
+wrap_cmpfunc(PyObject *self, PyObject *args, void *wrapped)
```
Comment (ltratt) [inline Objects/typeobject.c:6845]: Should this function emit a warning when it's called (to let the user know they're using something not supported in python3)?

Comment (nanjekyejoannah) [reply Objects/typeobject.c:6845]: Good one, I have done the needful in commit [3d24c44](https://github.com/softdevteam/pygrate3/pull/7/commits/3d24c44ae9e9872faa660542ba8a3fb899093108)

Comment (ltratt) [reply Objects/typeobject.c:6845]: Will that warn every time you call `cmp` in the same source location, or just one warning per location?

Comment (ltratt) [reply Objects/typeobject.c:6845]: Will that warn every time you call `cmp` in the same source location, or just one warning per location?

Comment (nanjekyejoannah) [reply Objects/typeobject.c:6845]: There are two areas to warn 1) Where cmp() is implemented i.e inside PyObject_Compare() and 2) At the __cmp__ implementation point where we attach the 'wrapfunc` logic to `TPSLOT('__cmp__'...`. So this was the last piece and only impacts `__cmp__` . The normal `cmp()` function already warns, the warning is implemented in:
`PyObject_Compare(PyObject *v, PyObject *w)`

**Comment by ltratt** [comment]
Please squash.

File: Objects/typeobject.c (hunk 1 above)
Comment (nanjekyejoannah) [reply Objects/typeobject.c:6845]: Formatting is wierd, bolded cmp == dunder "__cmp__"

**Comment by nanjekyejoannah** [comment]
Squashed.

**Pull request merged**